Feature/calypso reader blog tag preview

nbradbury opened the pull request:
Redesigns the headers shown in "blog preview" and "tag preview" to more closely match the Calypso web reader. Previously these headers were shown in the fragment, but now they're shown as custom view types in the reader RecyclerView.

[Demo video](https://cloudup.com/cjm9PJa3xQo)

Note: there may still be a few design tweaks made to this PR, but nothing that should require significant code changes.

tonyr59h commented:
Weird bug. If you select a tag->click Follow->click Following (so that you un-follow)->rotate the device the list won't refresh and shows "No posts with this tag"

nbradbury commented:
> Weird bug.

Oof. Looks like it also happens if you tap a tag you're following, tap Following to unfollow, then rotate. I'm looking into it.

tonyr59h commented:
[The tag text has some clipping](https://cldup.com/G3kyVnJrL4.png) on the bottom (g in bangalore).

tonyr59h commented:
With heavy use of the reader and an infinite back-stack I found it difficult to navigate back to the Masterbar. I ended up spamming back until the app backgrounded.

Definitely an edge case, but maybe there's something we can do to address it? I'm thinking an action bar item is overkill, we could do something like use the back button long press. Definitely not for this PR, I'll make a new issue if we decide to do something.

tonyr59h commented on WordPress/src/main/java/org/wordpress/android/ui/reader/views/ReaderBlogInfoView.java:
```diff
@@ -0,0 +1,181 @@
+        ReaderActions.ActionListener listener = new ReaderActions.ActionListener() {
+            @Override
+            public void onActionResult(boolean succeeded) {
+                if (!succeeded && getContext() != null) {
```
Maybe we should hold a reference to the listener and limit the requests to one at a time. Pressing it twice in succession can lead to contradictory toast messages. I'm assuming the same could happen on a slow connection.

tonyr59h commented:
Start in Reader Post->navigate to Reader Blog->change Follow setting->navigate back to Reader Post

The Follow status is not updated to reflect changes made in Reader Blog.

tonyr59h commented on WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderActivityLauncher.java:
```diff
@@ -99,25 +93,20 @@ public static void showReaderBlogPreview(Context context, ReaderPost post) {
         }
 
-        if (TextUtils.isEmpty(title)) {
-            title = ReaderBlogTable.getBlogNameFromFeedId(feedId);
```
The ReaderBlogTable import is unused with this removal.

tonyr59h commented on WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListActivity.java:
```diff
@@ -7,98 +7,64 @@
         if (actionBar != null) {
-            actionBar.setDisplayShowTitleEnabled(true);
-            actionBar.setDisplayHomeAsUpEnabled(true);
+            getSupportActionBar().setDisplayShowTitleEnabled(true);
```
nitpick; `actionBar` should be used here.

tonyr59h commented on WordPress/src/main/java/org/wordpress/android/ui/reader/adapters/ReaderPostAdapter.java:
```diff
@@ -155,22 +185,38 @@ public int getItemViewType(int position) {
     @Override
     public void onBindViewHolder(final RecyclerView.ViewHolder holder, final int position) {
-        if (holder instanceof SpacerViewHolder) {
+        if (!(holder instanceof ReaderPostViewHolder)) {
```
This reads weird to me but it seems like the best way to handle it since the majority of views will be `ReaderPostViewHolder`.

tonyr59h commented:
> Update: ... :) Fixed

:dancers:

nbradbury commented:
> The tag text has some clipping on the bottom

Fixed in a7edc55.

> The ReaderBlogTable import is unused with this removal

Fixed in 2238dcc.

> nitpick; actionBar should be used here.

D'oh! Fixed in 236ee51.

> Pressing [the follow button] twice in succession can lead to contradictory toast messages.

I chose to resolve this in 25c2f1f by disabling the follow button until the API call completes.

nbradbury commented:
> With heavy use of the reader and an infinite back-stack I found it difficult to navigate back to the Masterbar.

I'm also concerned about that, but I don't think we should add anything to the UI to enable navigating all the way back. For that to be helpful it would need to be obvious, and if it's obvious it's going to be extra noise for the majority of users who don't need it.

One thing I can do is hide the tag in the detail view when it's displayed from blog preview. That would prevent some "excessive" navigation, but that would be a limited solution since it wouldn't prevent deep navigation from tag preview.

nbradbury commented:
@tonyr59h Ready for round two!

tonyr59h commented:
> One thing I can do is hide the tag in the detail view when it's displayed from blog preview.

We should probably sit on this one and see if any users raise the issue.

tonyr59h commented:
Looks good!

:shipit:
